neqo-http3/src/connection_server.rs

```diff
@@ -216,24 +217,23 @@ impl Http3ServerHandler {
         buf: &mut [u8],
     ) -> Res<(usize, bool)> {
         qinfo!([self], "read_data from stream {}.", stream_id);
-        match self.base_handler.recv_streams.get_mut(&stream_id) {
-            None => {
-                self.close(conn, now, &Error::Internal);
-                Err(Error::Internal)
-            }
-            Some(recv_stream) => {
-                match recv_stream.read_data(conn, &mut self.base_handler.qpack_decoder, buf) {
-                    Ok((amount, fin)) => {
-                        if recv_stream.done() {
-                            self.base_handler.recv_streams.remove(&stream_id);
-                        }
-                        Ok((amount, fin))
-                    }
-                    Err(e) => {
-                        self.close(conn, now, &e);
-                        Err(e)
-                    }
+        let recv_stream = self
+            .base_handler
+            .recv_streams
+            .get_mut(&stream_id)
+            .ok_or(Error::InvalidStreamId)?;
+
+        let hs = recv_stream.http_stream().ok_or(Error::InvalidStreamId)?;
+        match hs.read_data(conn, buf) {
+            Ok((amount, fin)) => {
+                if recv_stream.done() {
```

Comment:
If you are able to use `hs.done()` here (I think that is possible), then you only need one variable.

Reply:
I should make HttpRecvStream extend RecvStream trait than it should work. Thanks.